src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/CompositeModelBinder.cs

```diff
@@ -76,12 +76,13 @@ public virtual Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindi
                             ValidationStateEntry entry;
                             if (!bindingContext.ValidationState.TryGetValue(result.Model, out entry))
                             {
-                                entry = new ValidationStateEntry();
+                                entry = new ValidationStateEntry()
+                                {
+                                    Key = result.Key,
+                                    Metadata = bindingContext.ModelMetadata,
+                                };
                                 bindingContext.ValidationState.Add(result.Model, entry);
                             }
-
-                            entry.Key = entry.Key ?? result.Key;
-                            entry.Metadata = entry.Metadata ?? bindingContext.ModelMetadata;
```

Comment:
This previously cleaned up incomplete `ValidationStateDictionary` entries, not just the one added just above. Not sure how important that was but it certainly compensated for odd 3rd party binders.